Workflow export ui

amyrebecca opened the pull request:
Adds a UI (a button on the data dumps page and a modal dialog) to export classifications for an individual workflow only, instead of an entire project at once.

# Review Checklist

- [x] Does it work in all major browsers: Firefox, Chrome, Edge, Safari?
- [x] Does it work on mobile?
- [x] Can you `rm -rf node_modules/ && npm install` and app works as expected?
- [x] Did you deploy a staging branch?
- [x] Did you get any type checking errors from [Babel Typecheck](https://github.com/codemix/babel-plugin-typecheck)


## Optional

- [x] If it's a new component, is it in ES6? Is it clear of warnings from ESLint?
- [x] Have you replaced any `ChangeListener` or `PromiseRenderer` components with code that updates component state?
- [x] If changes are made to the classifier, does the dev classifier still work?
- [ ] Have you added in [flow type annotations](https://flowtype.org/docs/type-annotations.html)?

camallen commented:
LGTM - tested and the UI works mostly. When i request a new workflow classification dump, i don't get an updated most recent link details for the existing / requested media (dump) resource.

![ui-w-export](https://cloud.githubusercontent.com/assets/295329/20388881/cb3c1936-acbe-11e6-9113-5d9017f3968f.jpg)

srallen commented on app/pages/lab/data-dumps.cjsx:
```diff
@@ -2,6 +2,8 @@ React = require 'react'
 counterpart = require 'counterpart'
 DataExportButton = require '../../partials/data-export-button'
 TalkDataExportButton = require '../../talk/data-export-button'
+ExportWorkflowsDialog = require('../../partials/export-workflows-dialog').default
```
Prefer ````import ExportWorkflowDialog from '../.../partials/export-workflows-dialog```` syntax with backticks, but not a blocking change.

srallen commented on app/partials/export-workflows-dialog.jsx:
```diff
@@ -0,0 +1,62 @@
+  }
+
+  componentDidMount(){
+    this.updateWorkflowsFromProject(this.props.project)
```
Could we add `defaultProps` and props validations for `this.props.project`?

srallen commented on app/partials/export-workflows-dialog.jsx:
```diff
@@ -0,0 +1,62 @@
+  }
+
+  requestDataExport(){
+    var workflow_id = this.refs.workflowList.value
```
Should be `const`

srallen commented on app/partials/export-workflows-dialog.jsx:
```diff
@@ -0,0 +1,62 @@
+    return (
+        <div>
+          <label className="form-label">Select a Workflow:</label>
+          <select size="5" ref="workflowList" className="multiline-select standard-input" style={{padding: "0.3vh 0.3vw"}} onChange={this.toggleExport.bind(this)}>
```
Binding to `this` should be done in the constructor. 

Also fyi, string literals in refs are being deprecated, so we should avoid this and find an alternative if possible.

srallen commented on app/partials/export-workflows-dialog.jsx:
```diff
@@ -0,0 +1,62 @@
+          </select>
+          <div style={{textAlign: "right"}}>
+            <button className="minor-button" style={{marginLeft: "1em"}}>Cancel</button>
+            <button className="standard-button" style={{marginLeft: "1em"}} disabled={!this.state.workflow_selected} onClick={this.requestDataExport.bind(this)}>Export</button>
```
Same comment as above about binding.

srallen commented:
Just noting that this closes #2944 when it merges.

amyrebecca commented:
I know I said I was out sick today but I wanted to get this in before the weekend. Anyways I'm not sure why my linter wasn't picking up the .eslintrc but since @srallen mentioned the linter errors I was able to get it fixed.

amyrebecca commented:
in re: @camallen's note, this is actually correct behavior--the link is for a full dump, whereas this generates only a partial dump

amyrebecca commented:
This should now be ready for merge. I should really know better than to test things with the Admin box checked, by the way...

camallen commented:
Not a blocker (can come in another PR) but the backend should create a dump media resource for the parital (per workflow) classification export and we can use this media resource for the links / timestamps for most recent as per [here](@zwolf )

@zwolf can you liaise with @amyrebecca to get these links working in the admin interface? This way people don't have to rely email links to get the data.

zwolf commented:
Related: https://github.com/zooniverse/Panoptes/pull/2063

amyrebecca commented:
Just discussed this with @zwolf and @trouille and the features that @camallen mentioned will be enabled by some pull requests against panoptes that aren't yet deployed. Laura was fine with merging this as-is and then circling back around to get the links on the next pass.

srallen commented on app/pages/lab/export-workflows-dialog.jsx:
```diff
@@ -27,10 +27,21 @@ class ExportWorkflowsDialog extends React.Component {
-        this.setState({ workflows });
+        let result = workflows;
+
+        if (order) {
```
`project.configuration.workflow_order` is an array and an empty array checked this way will be truthy. When `order` is an empty array, the `result` variable will set as an empty array and the dialog will not have workflows to select from.

amyrebecca replied:
This is actually quite tricky to accomplish since the export is started by clicking a button in a modal dialog. I'd like to merge this PR without that behavior unless you feel strongly about it.

srallen replied:
Yeah, there will need to be some refactoring to make it work. I'm ok with this being in a later iteration and PR.

amyrebecca replied:
Yeah absolutely--I have committed a fix for that.

amyrebecca commented:
I'm not sure where the comment went, but in re: the workflow export message, I'm not actually sure how to accomplish this because the export is performed from inside a modal dialog instead of by interacting with the main page. Maybe passing in an event handler to be fired?

srallen commented:
Looks like something maybe went wrong with a rebase. There are duplicate commits now from PRs that have already been merged into master, like the seven-ten stuff, help content, etc.

amyrebecca commented:
UGH yes, I see them. I'm not sure how they got in there, I'll have to try to remove them with a rebase if I can. I'm not sure why it shows as "okay to merge" under these circumstances

amyrebecca commented:
This is really confusing, the extraneous commits don't show up in my local branch. I may delete this pull request and recreate it?
<img width="749" alt="screen shot 2016-11-29 at 1 41 06 pm" src="https://cloud.githubusercontent.com/assets/8491638/20726079/b3cd4570-b639-11e6-8389-19870df39f9a.png">

amyrebecca commented:
!!!! now all those other commits are gone? but i did see them

srallen commented:
Hmm, maybe the issue was with github. Looks fine now. Weird. I'll do one last review locally just to be sure what I pull down looks good.

srallen commented:
I pulled down the latest and requested workflow classification exports for a couple of different test projects I have on staging yesterday afternoon. I still haven't received any emails for it, though if I request the normal classification export, I get that almost instantly since there are so few classifications. 

I know I said I wouldn't hold up for the export request state UI, but right now it just seems broken, so some kind of messaging seems necessary. I checked the network tab and it looks like the `POST` is being made successfully, but without a message or an email, I'm left without any information that it's working.

zwolf commented:
Emails on workflow exports are coming in a new Panoptes PR soon.

srallen commented:
@zwolf Thanks for the update. This PR should wait for that.

amyrebecca commented:
I definitely got emails containing CSV exports in the past, so I'm surprised this is broken now, especially given that the post means I didn't break it on our end, and the endpoint was working fine previously. I agree we should definitely wait until the PR is in and it's fine if we have to have the messaging, it'll just take one more commit.

zwolf commented:
For workflow classification exports? Staging or production? When?

amyrebecca commented:
i was getting them on staging pretty consistently during the time i was finishing up this PR, which was admittedly a couple of weeks ago now. the endpoint worked perfectly for me from the first time i called it. i suppose it could have been an artifact of which workflows i was exporting?

zwolf commented:
There appear to be two different paths by which classification export emails might be sent. 

The first is: `CreateClassificationsExport` (operation) --> `ClassificationsDumpWorker` --> `DumpMailerWorker`. This seems like the one that should be working for both.

Then there's this:`ClassificationDataMailerWorker` --> `ClassificationDataMailer`. Hm. I can't find anywhere that the former is actually called, so it must be some bit of magic. 

The second path is def what's sending the project version of this email and I'm not sure why the first is ignoring the `DumpMailerWorker` callback, cause it's clearly getting as far as the classifications dump worker. Nor why it seems to be intermittent, though it hasn't worked for me lately. Thoughts, @camallen / @marten ?

camallen commented:
@zwolf both paths use the same data mailer worker, the "magic" is [here](https://github.com/zooniverse/Panoptes/blob/7c0106054b20b4e13dc8e5f345784d8679b60e58/app/workers/concerns/dump_mailer_worker.rb#L14). Finally, seems the dump workers are stuck in retry loops on staging, https://panoptes-staging.zooniverse.org/sidekiq/retries

amyrebecca commented:
Added--as @srallen suggested--messaging to the user to indicate that their export data has been requested and that they'd receive it in their email (like the other export buttons do). Also added error reporting etc.

srallen commented on app/pages/lab/export-workflows-dialog.jsx:
```diff
@@ -0,0 +1,89 @@
+  }
+
+  requestDataExport() {
+    try {
```
It doesn't make sense to me to use a try/catch block instead of calling a `then()` and `catch()` on the api client promise. Is there a reason to use try/catch here. I think we would want to wait for the `POST` to complete and then catch the error or then perform a function.

This is what I'm describing to do instead: https://github.com/zooniverse/Panoptes-Front-End/blob/master/app/partials/data-export-button.cjsx#L23-L28

srallen commented:
I enabled the experimental tool flag for a test project of mine in staging and am seeing lots of errors. Not sure where it's coming from since the stack trace isn't super helpful:
![screen shot 2016-12-06 at 3 09 11 pm](https://cloud.githubusercontent.com/assets/5016731/20943662/1cccdf22-bbc6-11e6-8cff-5432a4d989d7.png)

Requesting for a workflow classification export also shows that error: 
<img width="681" alt="screen shot 2016-12-06 at 3 11 11 pm" src="https://cloud.githubusercontent.com/assets/5016731/20943700/40d84992-bbc6-11e6-984f-208a7c9770fe.png">

The UI for error messaging is looking good at least :)

amyrebecca commented:
I was getting that error intermittently too--I wonder if it's from the backend? I didn't change anything that would have affected the other promises that page fires (to get the links for previous exports)

zwolf commented:
@amyrebecca nope, it's not the backend: 
![screen shot 2016-12-06 at 3 33 34 pm](https://cloud.githubusercontent.com/assets/374835/20944546/80393a3a-bbc9-11e6-9bc3-76f3b9975d64.png)

amyrebecca commented:
You're right, I just ran the export on my machine and it worked fine.  This is strange.

amyrebecca commented:
It doesn't seem like any of these methods should throw 500s though?

<img width="711" alt="screen shot 2016-12-06 at 3 37 40 pm" src="https://cloud.githubusercontent.com/assets/8491638/20944747/3ad21434-bbca-11e6-9bda-159236a808c7.png">

zwolf commented:
If it's actually calling 
`          https://panoptes-                    staging.zooniverse.org/api/projects/1618/workflow_contents_export?admin=true:1staging.zooniverse.org/api/projects/1618`/workflow_contents_export?admin=true` 

then that'd def 404.

amyrebecca commented:
Oh uh, yeah, I would expect that not to work for sure, but let me look and see if that's the actual request being made or if the debug trace is just printing messy.

amyrebecca commented:
It's just bad pretty-printing, that request was actually:
<img width="361" alt="screen shot 2016-12-06 at 3 47 02 pm" src="https://cloud.githubusercontent.com/assets/8491638/20945074/656d06b2-bbcb-11e6-83b6-e4f30b0b4c78.png">

amyrebecca commented:
And 404s are fine, they actually return some JSON data and the client handles it, the real problem is when we get 500s instead
<img width="361" alt="screen shot 2016-12-06 at 3 47 44 pm" src="https://cloud.githubusercontent.com/assets/8491638/20945087/78e8c38e-bbcb-11e6-84be-adcf414804e8.png">

zwolf commented:
Yeah, it looks like something weird is going on. I'm looking into it.
